### ISSUE TITLE
Multiple Stickied Comments & DeltaLog Posts

### ISSUE THREAD

**Issue by Snorrrlax** [MystK/delta-bot-three#207]
I'm making a new issue for this since the [last one](https://github.com/MystK/delta-bot-three/issues/193) wasn't clear (though it's still worth reading for the comments and examples).

We currently have a bug where DeltaBot creates more than one [stickied comment](https://www.reddit.com/r/changemyview/comments/74dvaz/cmv_the_united_states_should_have_mandatory/dnxknos/) (and more than one [DeltaLog post](https://www.reddit.com/r/DeltaLog/comments/74e2cm/deltas_awarded_in_cmv_the_united_states_should/)) for the same CMV post. 

For those not familiar with this feature, DeltaBot is supposed to create just one stickied comment, which states the number of OP deltas and provides a link to the corresponding DeltaLog post containing every delta awarded. As you can imagine, this gets quite messy in /r/DeltaLog, and the multiple stickied comments cause confusion.

I'm not sure of the cause, but it doesn't always happen. Sometimes a stickied comment can stay in place and update with multiple OP deltas, and sometimes DeltaBot creates a new one for each OP delta.

**Comment by Snorrrlax** [comment]
@wyantb do you have an idea of what's going wrong here?

**Comment by wyantb** [comment]
My general assumption is that it's just a symptom of the normal synchronization issues...there's something like 4 or 5 different POST requests to reddit in order to get a post, sticky it, then make a new post based on the contents of the delta and save the information off in a wiki - if any POST in that flow gets interrupted, because of reddit flakes or we run out of allowed requests or whatever, then the submit will be in a partial state.

Only real options I can think of for addressing this: (1) attempt to scan all the comments in the post each time we award a delta, in order to determine if we've already made the comment; if so, attempt to continue the chain of POSTs where we left off.  This would probably increase our reads.  Or (2) attempt to make a different tradeoff here, by doing the POSTs in a different order; if we updated the wiki backing page asap - before stickying & mod-flagging the delta award comment - we could probably make it such that there'd never be a duplicated delta post (on the other hand, with the existing architecture, this would probably make it so that some posts aren't made to /r/DeltaLog).  

Either of those options sound appealing?

**Comment by Snorrrlax** [comment]
> this would probably make it so that some posts aren't made to /r/DeltaLog

In that case, I prefer the sound of your first option, but I don't know anything about how many reads we're allowed etc.

**Comment by nshiff** [comment]
"Clients connecting via OAuth2 may make up to 60 requests per minute."

https://github.com/reddit/reddit/wiki/API
linked from https://www.reddit.com/dev/api/

Depending on how frequently the bot runs, and how many requests it makes during each run, the 60 requests per minute may or may not be limiting.

**Comment by snugghash** [comment]
Can confirm that the bot does already include rate limiters based on looking at logs (correct me if I'm wrong). I'm looking over the code now, and will probably implement a solution similar to @wyantb 's option 1.

**Comment by snugghash** [comment]
Okay, here's what i believe is happening:
`findOrMakeStickiedComment` is called, which updates an existing sticky comment or makes one. I believe this is the only place in the entire codebase that this happens. 

We could add in code to scan all comments and retrieve an existing sticky from DeltaBot (currently it assumes this has been done before the function), and update if it exists within that function, right before it decides to make the comment.

If indeed the chain of POSTS has been interrupted, and if we want to solve this issue without adding such redundant code, we'd have to perform this check and restart the entire chain of requests. This is of course assuming each piece of code within the chain is already idempotent. If not, this is just too much work :(

**Comment by wyantb** [comment]
I would definitely not make the assumption that each part of the chain is idempotent - that's why we get duplicates in the first place (unless I'm misunderstanding you). DeltaBot can make top-level comments with identical contents - not idempotent. DB can make both sticky - not idempotent. DB can make duplicate posts to the log subreddit - not idempotent. Etc.

> We could add in code to scan all comments and retrieve an existing sticky from DeltaBot (currently it assumes this has been done before the function), and update if it exists within that function, right before it decides to make the comment.

Yip, this is what we want, to the best of my knowledge.

**Comment by Snorrrlax** [comment]
@MystK, can you assign @snugghash to this issue? I can't seem to do it.

**Comment by MystK** [comment]
It can only be assigned to a person if they have assign to deploy to production. Simply adding the in progress and mentioning that @snugghash is actively working on this might suffice.

**Comment by snugghash** [comment]
ETA Saturday. Can't really work on this right now :|

**Comment by Snorrrlax** [comment]
@snugghash are you still able to work on this?

**Comment by snugghash** [comment]
Yeah, sorry about the lack of updates. No work done yet, planning to do
this Friday night. If someone else wants to, they're welcome.

**Comment by snugghash** [comment]
Not going well. So much has come up, and this is so far down my list... I don't think I can get this done :(
Sorry. 
I'll push whatever progress I have into my fork.

**Comment by Snorrrlax** [comment]
Ah that's a shame, no worries though. Perhaps you could share some thoughts on how you were going to approach this, for anyone who might take it forward?

**Comment by snugghash** [comment]
@Snorrrlax Sure, but that's the major part of the work itself though.

Anyway, I was investigating, and found this:
`findOrMakeStickiedComment` relies on previous passing of the sticky comment's ID (its argument - `deltaLogPost.wikientry.stickiedCommentID`). There are two places `deltaLogPost.wikientry` is initialized other than `findOrMakeStickiedComment` - both in `findOrMakeDeltaLogPost` https://github.com/MystK/delta-bot-three/blob/e60c390d434cdaad5c56b5b6cdecbfdf4e67c404/src/index.js#L640.
The function gets the wikientry from a (possibly?) previous deltaLogPost, OR it makes a new one. While making a new one, https://github.com/MystK/delta-bot-three/blob/e60c390d434cdaad5c56b5b6cdecbfdf4e67c404/src/index.js#L670-L680 it never assigns the stickyCommentID itself (!). 

I believe it relies on a later updateDeltaLogPost call to update the contents with the stickyCommentID. So one possible way for this issue to occur is if that call fails - then no stickyCommentID is known, and another sticky comment is created. 

Have I understood what's going on correctly?

I propose we do the update again within `findOrMakeStickiedComment`, when we find no sticky comment.

Another way would be to search for sticky comment in all top level responses, not unlike https://github.com/MystK/delta-bot-three/blob/e60c390d434cdaad5c56b5b6cdecbfdf4e67c404/src/index.js#L705-L712  and update sticky comment content and stickyCommentID.

**Comment by wyantb** [comment]
> it never assigns the stickyCommentID itself (!).

> I believe it relies on a later updateDeltaLogPost call to update the contents with the stickyCommentID. So one possible way for this issue to occur is if that call fails - then no stickyCommentID is known, and another sticky comment is created.

> Have I understood what's going on correctly?

Yes, I believe your understanding is correct on all accounts. If we consider the primary block:

https://github.com/MystK/delta-bot-three/blob/e60c390d434cdaad5c56b5b6cdecbfdf4e67c404/src/index.js#L767-L769

One option could be to `updateDeltaLogWikiLinks` immediately after the deltalog post is made (this could help reduce the creation of duplicate /r/DeltaLog entries), IN ADDITION TO the current synchronization after the stickied comment is made. Anyway...your comment has inspired me to dig in a little bit further. BRB with a PR...

**Comment by wyantb** [comment]
OH DEAR LORD https://www.reddit.com/r/DeltaLog/wiki/revisions/internal HASN'T BEEN UPDATED IN 7 MONTHS. This is the wiki page the code uses to store links to the /r/DeltaLog posts it has made...eek. Has it hit the max wiki size limit? It's all > 6 months anyway, reddit archives content after that point right? I think we should axe all the content of the whole wiki page.

**Comment by wyantb** [comment]
Actually...who's got logs to the bot? https://www.reddit.com/r/DeltaLog/wiki/internal only lists a content-length of some 128 KB - is that already hitting the limits? Hrm. 

For now I've cleared the content of the wiki page and taken a backup of the JSON that was within the param start/end tags:
[reddit-wiki-contents.json.txt](https://github.com/MystK/delta-bot-three/files/1485685/reddit-wiki-contents.json.txt)

If that helps, then we were definitely hitting the max length, and we'll probably want to come up with some other solution because only ~3 months of deltalogs were able to get saved to the wiki page before updates stopped happening.

**Comment by wyantb** [comment]
Hah, sweet. I'm seeing good stuff in https://www.reddit.com/r/changemyview/comments/7e31g5 so clearing the deltalog internal wiki page should've gone a loooooooong way to help with this.

Still, this problem supports the idea of us having our own data store. Reddit capping what we can put in a wiki at a measly 128 KB certainly sounds to me like we should get a bigger box to run this thing on and have a data store (I mentioned Scaleway in another thread as a potential provider)

**Comment by MystK** [comment]
Ya looks like we'll need to move to a database
________________________________

**Comment by Snorrrlax** [comment]
@hacksoncode what are your thoughts on this?

**Comment by hacksoncode** [comment]
How big an instance do we need? I'm ok with paying for a somewhat larger one... 

Everything would have to be migrated, because apparently you can't just "upgrade" a 128MB instance at RamNode (though maybe they'd do it if we asked nicely, I don't know)... but that would true elsewhere too.

**Comment by Snorrrlax** [comment]
Side note: Is there a way for any willing mods to add to a sort of "money pool" that pays for this? I wouldn't mind chipping in - you shouldn't have to cover it yourself. Maybe there could even be a donation button for our users if it gets more expensive.

**Comment by hacksoncode** [comment]
At the present time that's not necessary, but I'll let you know.

**Comment by wyantb** [comment]
I doubt I know for sure, though I'm guessing a `2GB SVZS` would do the trick for $7/month or a Scaleway `C1` (https://www.scaleway.com/) 3EUR/month. 

On a DB implementation: https://github.com/level/levelup looks like a pretty dang simple key value store, backed by some pretty good heavyweights. `yarn add levelup leveldown` and their code snippets look very straightforward - I just ran that yarn command, followed their example, and it pumped out a quick `mydb` folder to the local dir. `ncdu` reports a total of 90MB of disk space taken up by `node_modules` after including these deps, so certainly in terms of code impact that's not very large.

Their features look pretty hot too:

```
Multiple changes can be made in one atomic batch.
Users can create a transient snapshot to get a consistent view of data.
Forward and backward iteration is supported over the data.
Data is automatically compressed using the Snappy compression library.
```

So, to wrap up with a meaningful conclusions here, for a better estimate about how big a box we'd need, I think we'd want to analyze what we plan on storing on disk. If we're just talking about the DeltaLogs, 3 months of history was only 128KB, so just about any option should work for us. If we're also planning on taking a dump of the entire subreddit's history and keeping that on box, that will dominate our storage needs.

**Comment by hacksoncode** [comment]
I'm ok with either of those options... (the C1 service at scaleway seems to be ARM based, but it looks like they have a similar X64 service with basically the same stats for the same price). 

What would be the best way to migrate?

Anyway, you guys let me know what you decide and I'll take care of setting up the server instance...

**Comment by Snorrrlax** [comment]
@wyantb and @MystK: Thoughts on how to migrate? Also, is doing this & merging wyantb's PR a likely end to this problem?

**Comment by wyantb** [comment]
Ah, sorry for slow response time - busy as ever.

For migrating, zipping the whole directory of the DeltaBot wherever it's deployed - minus `node_modules` - should do the trick, then you could `yarn install` the deps back into place on the target machine.

The largest source of this bug, though, by far should have been the inability to edit the internal wiki page. With that addressed, we should have a month or so of relatively stable/error-free operation...let me know if that's not the case. The PR (if I wrote it correctly) would also help. That said if I get around to doing the whole DB thing soon enough, it might be worth testing both in one pass. Depends on how y'all want to handle the staging environment.

tl;dr: migration is only useful once we have the DB in place. We don't have that yet. In the meantime, consider testing out my PR + the editable wiki page; things should be much better than they were the past few months.

**Comment by Snorrrlax** [comment]
@hacksoncode, thoughts?

**Comment by wyantb** [comment]
OK, I updated the PR associated here, see https://github.com/MystK/delta-bot-three/pull/222 . This now uses `levelup` (backed by a `leveldown` db) to store the deltalog link information, rather than the old `internal` wiki page. We'll definitely want to get it merged and deployed within a month or two so that we don't run out of space on the internal wiki and have the deltalog stuff go haywire again.

@hacksoncode need anything in addition to my prior comments in https://github.com/MystK/delta-bot-three/issues/207#issuecomment-346520426 in terms of migration to a new instance? Do you want to setup the bot on a staging instance yourself first to try it out before merging my PR and deploying to a new, larger box?

**Comment by Snorrrlax** [comment]
@wyantb is this permanently fixed now, or was a temporary measure used?

**Comment by wyantb** [comment]
As long as the box doesn't run out of disk space, the wiki space issue is permanently worked around.

The separate general principle - that we can't atomically issue a batch of commands against reddit API servers, such that if some intermediate steps fails its hard to recover from - wasn't tackled. I don't know how much of an issue this is in practice.

So...how's it been behaving for the past 4 weeks, apart from https://github.com/MystK/delta-bot-three/issues/227 ? The evidence is all in the behavior, not my head.

**Comment by MystK** [comment]
@wyantb do you think this is safe to close?
